<!doctype html>
<html>
  <head>
    <meta charset="utf-8">
    <meta http-equiv="X-UA-Compatible" content="chrome=1">
    <title>Baa.pl by josephryan</title>

    <link rel="stylesheet" href="stylesheets/styles.css">
    <link rel="stylesheet" href="stylesheets/pygment_trac.css">
    <meta name="viewport" content="width=device-width, initial-scale=1, user-scalable=no">
    <!--[if lt IE 9]>
    <script src="//html5shiv.googlecode.com/svn/trunk/html5.js"></script>
    <![endif]-->
  </head>
  <body>
    <div class="wrapper">
      <header>
        <h1>Baa.pl</h1>
        <p>use BLAT to ASSESS an ASSEMBLY <p>
This program parses the output of a BLAT run of transcriptome vs. a genome. It returns 1) Total % coverage, 2) Average number of contigs mapped per mapped transcript, and 3) number of potential rearrangements. It will also create a file with transcript IDs that occur at regions that are potentially misassembled. It is simple to run. Just run the blat (transcripts vs. genome) and then run the script with the blat-output and your transcript-fasta-file as arguments.</p>

        <p class="view"><a href="https://github.com/josephryan/baa.pl">View the Project on GitHub <small>josephryan/baa.pl</small></a></p>


        <ul>
          <li><a href="https://github.com/josephryan/baa.pl/zipball/master">Download <strong>ZIP File</strong></a></li>
          <li><a href="https://github.com/josephryan/baa.pl/tarball/master">Download <strong>TAR Ball</strong></a></li>
          <li><a href="https://github.com/josephryan/baa.pl">View On <strong>GitHub</strong></a></li>
        </ul>
      </header>
      <section>
        <h3>Welcome to the baa.pl GitHub Page.</h3>

<p>There are two main uses of this script. 
<ol>
<li>It is useful to use it to compare alternative assemblies. Maximizing how well transcripts align to an assembly is an independent criteria that can be used along with N50 and other stats to objectively choose between assemblies.
<li>It can be used to measure the completeness of an assembly in terms of coding sequence. This can be especially useful if you have an assembly that has lots of collapsed repeats so the assembly sizes are smaller than the estimated genome size.
</ol>
A typical session with blat and baa.pl goes like this:
</p>
<pre><code>$ blat genome_assembly1.fa ests.fa ests_v_genome_assembly1.blat
$ baa.pl ests_v_genome_assembly1.blat ests.fa > ests_v_genome_assembly1.baa
</code></pre>

<h3>Authors and Contributors</h3>

<p>Joseph F. Ryan, Ph.D. (<a href="https://github.com/josephryan" class="user-mention">@josephryan</a>)</p>

<h3>Support or Contact</h3>

<p>There is <a href="https://github.com/josephryan/baa.pl/wiki/Documentation">Documentation</a> on the wiki.</p>

<p>If you have a negative or positive experience with baa.pl, I am interested in hearing about it. E-Mail: josephryan@yahoo.com or leave something on the <a href="https://github.com/josephryan/baa.pl/wiki">wiki</a>.</p>
      </section>
      <footer>
        <p>This project is maintained by <a href="https://github.com/josephryan">josephryan</a></p>
        <p><small>Hosted on GitHub Pages &mdash; Theme by <a href="https://github.com/orderedlist">orderedlist</a></small></p>
      </footer>
    </div>
    <script src="javascripts/scale.fix.js"></script>
    
  </body>
</html>
